Horizontal Scrolling

bambu opened the pull request:
Added support for horizontal scrolling. The code comes from the original
vim source. Removed the check for `guioptions` as Neovim does not have
support for that setting. Including a test for `<ScrollWheelLeft>` and
`<ScrollWheelRight>` in `mouse_spec.lua`.

This PR is for #3440.

Commit:
Horizontal Scrolling
Added support for horizontal scrolling. The code comes from the original
vim source. Removed the check for `guioptions` as Neovim does not have
support for that setting. Including a test for `<ScrollWheelLeft>` and
`<ScrollWheelRight>` in `mouse_spec.lua`.

justinmk commented:
Lint errors. Also, functions without parameters require (void).

ianks commented on src/nvim/normal.c:
```diff
@@ -3611,6 +3625,107 @@ void scroll_redraw(int up, long count)
+    col = 0;
+    if (*p != NUL)
+      for (;;)
+      {
```
This if statement is very ambiguous

bambu replied:
That is the way that it is written in the VIM source. If necessary, I can modify it to be less ambiguous but I tried to leave the code as close as possible to the original.

ZyX-I replied:
@bambu “As close as possible” is a bad idea. We have different code style.

ZyX-I replied:
`{` should be on the same line as `for`.

Commit:
Fixed linting issues. Re-ordered functions

bambu commented:
Is there any action needed that is preventing this from being merged?

ZyX-I commented on src/nvim/edit.c:
```diff
@@ -7620,6 +7620,15 @@ static void ins_mousescroll(int dir)
       else
         scroll_redraw(dir, 3L);
+    } else {
+      int val, step = 6;
```
Split the definition.

ZyX-I commented on src/nvim/edit.c:
```diff
@@ -7620,6 +7620,15 @@ static void ins_mousescroll(int dir)
+      int val, step = 6;
+
+      if (mod_mask & (MOD_MASK_SHIFT | MOD_MASK_CTRL))
+        step = curwin->w_width;
```
`{}` over each `if` body.

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
 }
 
 /*
```
Multiline `/**/` comments are not allowed, use `///`. How did linter accept this?

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
+     * line numbers, topline and botline can be invalid when displaying is
+     * postponed. */
+    if (curwin->w_topline <= curwin->w_cursor.lnum
+      && curwin->w_botline > curwin->w_cursor.lnum
```
Wrong indentation. `&&` should be on the same column as the very first `curwin`.

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
+     * postponed. */
+    if (curwin->w_topline <= curwin->w_cursor.lnum
+      && curwin->w_botline > curwin->w_cursor.lnum
+      && curwin->w_botline <= curbuf->b_ml.ml_line_count + 1)
```
`{` should be on the same line as `if`, or at least the last `)`

ZyX-I commented on src/nvim/edit.c:
```diff
@@ -7620,6 +7620,15 @@ static void ins_mousescroll(int dir)
+        step = curwin->w_width;
+      val = curwin->w_leftcol + (dir == MSCR_RIGHT ? -step : step);
+      if (val < 0)
+        val = 0;
```
`{}`

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3538,6 +3538,20 @@ static void nv_mousescroll(cmdarg_T *cap)
+  } else {
+    /* Horizontal scroll - only allowed when 'wrap' is disabled */
+    if (!curwin->w_p_wrap) {
+      int val, step = 6;
```
Split.

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3538,6 +3538,20 @@ static void nv_mousescroll(cmdarg_T *cap)
+      int val, step = 6;
+
+      if (mod_mask & (MOD_MASK_SHIFT | MOD_MASK_CTRL))
+          step = curwin->w_width;
```
{}

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3538,6 +3538,20 @@ static void nv_mousescroll(cmdarg_T *cap)
+
+      val = curwin->w_leftcol + (cap->arg == MSCR_RIGHT ? -step : +step);
+      if (val < 0)
+          val = 0;
```
{}

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
+static colnr_T scroll_line_len(linenr_T lnum)
+{
+    char_u  *p;
+    colnr_T col;
```
Should be moved to the place where col is set to zero.

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
+ */
+static colnr_T scroll_line_len(linenr_T lnum)
+{
+    char_u  *p;
```
Should be moved to the place where `p` is assigned.

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
+
+    p = ml_get(lnum);
+    col = 0;
+    if (*p != NUL)
```
{}

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
+      {
+          w = chartabsize(p, col);
+          mb_ptr_adv(p);
+          if (*p == NUL)    /* don't count the last character */
```
{}, `//` comment

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
+/*
+ * Find longest visible line number.  If this is not possible (or not desired,
+ * by setting 'h' in "guioptions") then the current line number is returned.
+ */
```
`///`

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
+
+    /* Calculate maximum for horizontal scrollbar.  Check for reasonable
+     * line numbers, topline and botline can be invalid when displaying is
+     * postponed. */
```
`//`

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
+ */
+static linenr_T gui_find_longest_lnum(void)
+{
+    linenr_T ret = 0;
```
Wrong indentation: should be two spaces.

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
+{
+    char_u  *p;
+    colnr_T col;
+    int   w;
```
Move to where `w` is assigned. This is absolutely not needed in the function scope.

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
+    colnr_T col;
+    int   w;
+
+    p = ml_get(lnum);
```
Wrong indentation.

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
+       * below. */
+      for (lnum = curwin->w_topline; lnum < curwin->w_botline; ++lnum)
+      {
+          n = scroll_line_len(lnum);
```
Wrong indent.

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
+      /* Use maximum of all visible lines.  Remember the lnum of the
+       * longest line, closest to the cursor line.  Used when scrolling
+       * below. */
+      for (lnum = curwin->w_topline; lnum < curwin->w_botline; ++lnum)
```
Move `{` here.

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
+      colnr_T     n;
+      long      max = 0;
+
+      /* Use maximum of all visible lines.  Remember the lnum of the
```
`//`

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
+      && curwin->w_botline > curwin->w_cursor.lnum
+      && curwin->w_botline <= curbuf->b_ml.ml_line_count + 1)
+    {
+      linenr_T    lnum;
```
Should be `for (linenr_T lnum =…`, not here.

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
+      && curwin->w_botline <= curbuf->b_ml.ml_line_count + 1)
+    {
+      linenr_T    lnum;
+      colnr_T     n;
```
Should be inside the `for` cycle.

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
+            max = n;
+            ret = lnum;
+          } else if (n == (colnr_T)max
+            && abs((int)(lnum - curwin->w_cursor.lnum))
```
Wrong indent.

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
+          }
+      }
+    } else {
+      /* Use cursor line only. */
```
`//`

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
+static linenr_T longest_lnum;
+/*
+ * Do a horizontal scroll.  Return TRUE if the cursor moved, FALSE otherwise.
+ */
```
///

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
+ */
+int gui_do_horiz_scroll(int leftcol, bool compute_longest_lnum)
+{
+    /* no wrapping, no scrolling */
```
//

ZyX-I replied:
Wrong indent: 4 in place of 2.

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
+{
+    /* no wrapping, no scrolling */
+    if (curwin->w_p_wrap)
+      return false;
```
{}

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
+      return false;
+
+    if (curwin->w_leftcol == (colnr_T)leftcol)
+      return false;
```
{}

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
+    /* When the line of the cursor is too short, move the cursor to the
+     * longest visible line. */
+    if (!virtual_active()
+      && (colnr_T)leftcol > scroll_line_len(curwin->w_cursor.lnum))
```
Align after (

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
+    if (!virtual_active()
+      && (colnr_T)leftcol > scroll_line_len(curwin->w_cursor.lnum))
+    {
+      if (compute_longest_lnum)
```
if(){

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
+          curwin->w_cursor.lnum = gui_find_longest_lnum();
+          curwin->w_cursor.col = 0;
+      } else if (longest_lnum >= curwin->w_topline
+        && longest_lnum < curwin->w_botline) {
```
Align

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
+          curwin->w_cursor.col = 0;
+      } else if (longest_lnum >= curwin->w_topline
+        && longest_lnum < curwin->w_botline) {
+        /* Do a sanity check on "longest_lnum", just in case. */
```
//

ZyX-I commented on src/nvim/normal.c:
```diff
@@ -3612,6 +3626,108 @@ void scroll_redraw(int up, long count)
+    if (*p != NUL)
+      for (;;)
+      {
+          w = chartabsize(p, col);
```
Wrong indent as far as I see.

ZyX-I commented:
Linter is close to useless here.

Commit:
Fixed style to meet neovim

Commit:
Merge branch 'master' of https://github.com/neovim/neovim into horizontal-scroll

christian-schulze commented:
Dump.

This would be a really awesome addition to neovim.

justinmk commented on src/nvim/normal.c:
```diff
@@ -3940,6 +3957,101 @@ void scroll_redraw(int up, long count)
+///
+/// Do a horizontal scroll.  Return TRUE if the cursor moved, FALSE otherwise.
+///
+int gui_do_horiz_scroll(int leftcol, bool compute_longest_lnum)
```
This and the other 2 new functions can be moved to `mouse.c` after adding these includes to mouse.c:

    #include "nvim/memline.h"
    #include "nvim/charset.h"

justinmk commented on src/nvim/normal.c:
```diff
@@ -3940,6 +3957,101 @@ void scroll_redraw(int up, long count)
+
+///
+/// Find longest visible line number.  If this is not possible (or not desired,
+/// by setting 'h' in "guioptions") then the current line number is returned.
```
You removed the check for 'guioptions' so let's remove that sentence from this comment. Also makes sense to remove the `gui_` prefix from the function name now.

justinmk commented on src/nvim/normal.c:
```diff
@@ -3867,6 +3867,23 @@ static void nv_mousescroll(cmdarg_T *cap)
+      }
+
+      gui_do_horiz_scroll(val, true);
+    }
```
This whole block is identical with the block added in edit.c. Can you move it to a static function in mouse.c, something like `mouse_scroll_horiz`. Then `gui_do_horiz_scroll` can be made static, too.

Commit:
Merge remote-tracking branch 'upstream/master' into horizontal-scroll

Commit:
Moved horizontal scrolling code to mouse.c as requested.
Removed 'gui_' from the function names, as well as the
guioptions comment. Renamed variables to be more specific
(as opposed to generic p, w). Made single use functions static

bambu commented:
@justinmk Made the requested updates.

justinmk commented:
@bambu LGTM, thanks for followup! Merged w/ squash.
